naga/src/proc/constant_evaluator.rs

```diff
@@ -862,6 +888,86 @@ impl<'a> ConstantEvaluator<'a> {
             crate::MathFunction::Cosh => {
                 component_wise_float!(self, span, [arg], |e| { Ok([e.cosh()]) })
             }
+            crate::MathFunction::CountLeadingZeros => {
+                component_wise_concrete_int!(self, span, [arg], |e| {
+                    #[allow(clippy::useless_conversion)]
+                    Ok([e
+                        .leading_zeros()
+                        .try_into()
+                        .expect("bit count overflowed 32 bits, somehow!?")])
+                })
+            }
+            crate::MathFunction::CountOneBits => {
+                component_wise_concrete_int!(self, span, [arg], |e| {
+                    #[allow(clippy::useless_conversion)]
+                    Ok([e
+                        .count_ones()
+                        .try_into()
+                        .expect("bit count overflowed 32 bits, somehow!?")])
+                })
+            }
+            crate::MathFunction::CountTrailingZeros => {
+                component_wise_concrete_int!(self, span, [arg], |e| {
+                    #[allow(clippy::useless_conversion)]
+                    Ok([e
+                        .trailing_zeros()
+                        .try_into()
+                        .expect("bit count overflowed 32 bits, somehow!?")])
+                })
+            }
+            crate::MathFunction::Degrees => {
+                component_wise_float!(self, span, [arg], |e| { Ok([e.to_degrees()]) })
+            }
+            crate::MathFunction::Exp => {
+                component_wise_float!(self, span, [arg], |e| { Ok([e.exp()]) })
+            }
+            crate::MathFunction::Exp2 => {
+                component_wise_float!(self, span, [arg], |e| { Ok([e.exp2()]) })
+            }
+            crate::MathFunction::Floor => {
+                component_wise_float!(self, span, [arg], |e| { Ok([e.floor()]) })
+            }
+            crate::MathFunction::Fract => {
+                component_wise_float!(self, span, [arg], |e| { Ok([e.fract()]) })
```

Comment:
```suggestion
                component_wise_float!(self, span, [arg], |e| { Ok([e - e.trunc()]) })
```
Rust's `.fract()` impl seems to be `e - e.floor()` but the WGSL spec says it should be `e - e.trunc()`.

Reply:
uh

The source for [`f32::fract`](https://doc.rust-lang.org/stable/src/std/f32.rs.html#165-167) and [`f64::fract`](https://doc.rust-lang.org/stable/src/std/f64.rs.html#165-167) both contain `self - self.trunc()` as the body. So, your problem statement doesn't seem valid.

I feel this is conclusively resolved, but I'll let you hit `Resolve` just in case.

Reply:
Sorry, I mistyped the comment. Rust does use the truncated version but the WGSL spec expects the floored version (i.e. the opposite of my previous comment).

Reply:
uh

I don't think that that's right, either. In the WGSL spec. (and other places AFAIK), `trunc` goes _towards 0_ on the number scale. IOW, that means `floor`ing when positive, and `ceil`ing when negative. This is what the Rust's `trunc` implements, too, i.e., [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7a8316d77509c504ba2d60bb54df428f):

```rust
assert_eq!(1.25f32.trunc(), 1.0);
assert_eq!(-3.14f32.trunc(), -3.0); 
```

From [the spec. text on `truncate`](https://gpuweb.github.io/gpuweb/wgsl/#truncate):

> The truncate function is defined over real numbers x extended with +∞ and −∞:
> 
> -   truncate(+∞) = +∞
>     
> -   truncate(−∞) = −∞
>     
> -   for real number x, computes the nearest whole number whose absolute value is less than or equal to the absolute value of x:
>     
>     -   truncate(x) = ⌊x⌋ if x ≥ 0, and ⌈x⌉ if x < 0.

Meanwhile, if we were to always `floor`, the spec. would say, `truncate(x) = ⌊x⌋` with no condition noted. If written as Rust (rather than WGSL), that would mean this shouldn't panic:

```rust
assert_eq!(1.25f32.trunc(), 1.0);
assert_eq!(-3.14f32.trunc(), -4.0); // RHS changed
```

Reply:
oderp, I got lost looking at the definition of the `truncate` logic, and forgot to connect it back to the actual matter at hand (which is `fract`). You're right, and I'm a ditz. Fixing now!

Reply:
Fixed up to be `e - e.floor()` with 7e78aca19. Sorry, I know it took me long enough to process what you were saying! Thanks for catching this. 😀💦

Reply:
No worries, thanks for addressing all the feedback! :)